modules/distributor/distributor.go

```diff
@@ -575,16 +584,32 @@ func requestsByTraceID(batches []*v1.ResourceSpans, userID string, spanCount int
 	return keys, traces, nil
 }
 
-func countDiscaredSpans(numSuccessByTraceIndex []int, lastErrorReasonByTraceIndex []tempopb.PushErrorReason, traces []*rebatchedTrace, repFactor int) (maxLiveDiscardedCount, traceTooLargeDiscardedCount, unknownErrorCount int) {
+func discardedTraces(numSuccessByTraceIndex []int, lastErrorReasonByTraceIndex []tempopb.PushErrorReason, traces []*rebatchedTrace, repFactor int) []*discardedTrace {
```

Comment:
so i like how you tried to organize this, but we can not incur the allocations of these discardedTraces structs on our write path.

even tho it may be ugly, we're just going to have to work the logging of the dropped spans into here directly:

https://github.com/grafana/tempo/blob/main/modules/distributor/distributor.go#L584

Reply:
Could you explain your concern a bit more? Is it about the memory footprint? `unsafe.Sizeof(discardedTrace{})` is just 16 bytes (one pointer and one enum). So, I guess it would be barely measurable, as these are just allocated for discarded spans. On the normal write path, there is no additional allocation when no spans are discarded. Even if all spans are discarded, the footprint remains low from my perspective. Or is it a different concern?

Reply:
sure, it's not so much about the number of bytes allocated. i'm more concerned with the allocations of the slice and the individual struct pointers. creating this many objects on the heap would put undue pressure on the GC.

in a situation where the ingesters were struggling and started dropping more spans this would create a potential cascading failure where the distributors started creating a large number of objects, falling behind on heap cleanup and OOMing.

since this is a "debug" option i expect additional work/allocations when its turned on, but would prefer it have no impact when turned off.

Reply:
I have to admit that I am not fully convinced that allocating or resizing a slice (sub millisecond time) is something that would hav any measurable impact compared to any network request (at least milliseconds). 
However, I went thought the code again and I think I found a good solution now, which shouldn't have any runtime impact and still is not too ugly to contribute it. 

I also adjusted the trace logging a bit (added `push_error_reason` and `tenant`), which might be helpful. Adding the tenant could also be something to consider for `log_received_spans`. 

Please do a fresh review on these changes.